<font size="+3"><b>Assignment 2: Linear Models and Validation Metrics</b></font>

***
* **Full Name** = Nathan Ante
* **UCID** = 30157706
***

<font color='Blue'>In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.</font>

You can use the Table of Content on the left side of this notebook to efficiently navigate within this documents.

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|           **Part 2: Regression**           |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    0.5    |
| Step 3: Implement Machine Learning   Model |     1     |
|            Step 4: Validate Mode           |     1     |
|          Step 5: Visualize Results         |     1     |
|                  Questions                 |     2     |
|             Process Description            |     4     |
|  **Part 3:   Observations/Interpretation** |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **30**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **4**   |

# **Part 1: Classification (14.5 marks total)**

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|                  **Total**                 |  **14.5** |

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:


## **Step 0:** Import Libraries

In [1]:
import numpy as np
import pandas as pd

C:\Users\MacKante\AppData\Local\Temp\ipykernel_2096\1662815981.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [2]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_spam

# TO DO: Print size and type of X and y
X, y = load_spam()

print(f'Size of X: {len(X)} | Type of X: {type(X)}')
print(f'Size of Y: {len(y)} | Type of Y: {type(y)}')

Size of X: 4600 | Type of X: <class 'pandas.core.frame.DataFrame'>
Size of Y: 4600 | Type of Y: <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [3]:
# TO DO: Check if there are any missing values and fill them in if necessary
missing_X = X.isnull().sum()
missing_y = y.isnull().sum()

# Check if there is anything missing in X, if there is then fill it with the mean of X
if missing_X.any():
    print(f"There are {missing_X.sum()} missing values in X, filling them with the mean")
    X.fillna(X.mean())
else:
    print("There are no missing values in X")

There are no missing values in X


For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [4]:
# TO DO: Create X_small and y_small

# Adding imports
from sklearn.model_selection import train_test_split

# Use 5% for the test set
X_small, _, y_small, _ = train_test_split(X, y, train_size=0.05, random_state=42)

# Print lengths
print("Size of X_small:", X_small.shape)
print("Size of y_small:", len(y_small))

Size of X_small: (230, 57)
Size of y_small: 230


## **Step 3:** Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

## **Step 4:** Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

## **Step 5:** Visualize Results (4 marks for steps 3-5)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [5]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

# imports
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Dataframe for results
results = pd.DataFrame(columns=["Data Size", "Training Accuracy", "Validation Accuracy"])

# Create a list of the datasets and run it through a loop
# Index 0: X and y dataset
# Index 1: First two columns of X and y
# Index 2: X_small and y_small

X_arguments = [X, X.iloc[:, :2], X_small]
y_arguments = [y, y, y_small]
indexes = ['Full Dataset', 'First Two Columns', 'Small Dataset']

for i in range(3):
    # Instantiate model
    model = LogisticRegression(max_iter=2000)

    # Split training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X_arguments[i], y_arguments[i], test_size=0.2, random_state=0)

    # Implement model
    model.fit(X_train, y_train)

    # Training accuracy
    training_predictions = model.predict(X_train)
    train_accuracy = accuracy_score(y_train, training_predictions)

    # Validation accuracy
    validation_predictions = model.predict(X_test)
    validation_accuracy = accuracy_score(y_test, validation_predictions)
    
    results.loc[f'{indexes[i]}'] = [len(X_arguments[i]), train_accuracy, validation_accuracy]

# Print the results dataframe
print(results)


                   Data Size  Training Accuracy  Validation Accuracy
Full Dataset          4600.0           0.927446             0.938043
First Two Columns     4600.0           0.614946             0.593478
Small Dataset          230.0           0.961957             0.891304


## **Questions (4 marks)**
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

<font color='Green'><b>YOUR ANSWERS HERE</b></font>

### Results
| Dataset            | Data Size    | Training Accuracy | Validation Accuracy |
|--------------------|--------------|-------------------|---------------------|
| Full dataset       | 4600.0       | 0.927446          | 0.938043            |
| First two columns  | 4600.0       | 0.614946          | 0.593478            |
| Small dataset      | 230.0        | 0.961957          | 0.891304            |


1. The training and validation accuracy could change greatly depending on the amount of data used. The full dataset produces high accuracy with a very small difference between them (about 0.011 difference and favoring validation accuracy). The model with only the first two columns performed the worst, with about ~0.3 difference in accuracy compared to the other two models. The smaller dataset with 5% of the size of the full dataset performed similarly to the full dataset, although there is a greater difference between their accuracies compared to the full dataset (about 0.07 difference and favoring training accuracy). In conclusion, it shows that using a more full/complete dataset leads to higher than using a less complete one, especially with less columns. In this case, the small dataset and full dataset were similar in values, but using more entries would result in a more accurate representation of the data.

2. According to the documentation, the Spam dataset is about spam. In this context, a false positive would represent emails that were actually real emails were classified as spam by the program. A false negative would represent emails that were spam were classified as real emails. Depending on the person both could be bad, if its a false positive, you could lose important emails in spam, if its a false negative, youre getting an inbox full of spam. So you either lose actual emails or soldier through a mountain of spam, personally I would choose false positive.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>

1. Where did you source your code?

    - I mainly used the Lab assignment source code as a guideline on the majority of the code, specifically labs 2 and 3. I also reviewed some of the slides on d2l, specifically on the data processing and model selection slides.

2. In what order did you complete the steps?

    - I followed the steps in the notebook.

3. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?

    - I used generative AI to teach me how to explain how some of the lab source code worked, mainly so I could learn how it related to the validation metrics I had to calculate. I also asked how to do some of the imports and I also asked it to make me a markdown table of the results I calculated. I also asked it how other basic python functionalities worked as I am not well versed on this language compared to others.

4. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

    - Admittedly I am not the best at python, I am not as good at remembering its functionalities and methods and constantly have to look up even some of the basic stuff. But generative AI and and some of the lab source code has made it easier to write code for the assignment. I also got lost around steps 3-5 a little so I had to refactor my code a couple of times in order to make it cleaner, instead of a loop I used to manually make a model and calculations for each dataset.

# **Part 2: Regression (10.5 marks total)**

| **Question**                               | **Point** |
|--------------------------------------------|-----------|
| **Part 2: Regression**                     |           |
| Step 1: Data Input                         | 1         |
| Step 2: Data Processing                    | 0.5       |
| Step 3: Implement Machine Learning   Model | 1         |
| Step 4: Validate Mode                      | 1         |
| Step 5: Visualize Results                  | 1         |
| Questions                                  | 2         |
| Process Description                        | 4         |
| **Total**                                  | **10.5**  |

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [6]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_concrete

# TO DO: Print size and type of X and y
X, y = load_concrete()

print(f'Size of X: {len(X)} | Type of X: {type(X)}')
print(f'Size of Y: {len(y)} | Type of Y: {type(y)}')

Size of X: 1030 | Type of X: <class 'pandas.core.frame.DataFrame'>
Size of Y: 1030 | Type of Y: <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [7]:
# TO DO: Check if there are any missing values and fill them in if necessary
missing_X = X.isnull().sum()
missing_y = y.isnull().sum()

# Check if there is anything missing in X, if there is then fill it with the mean of X
if missing_X.any():
    print(f"There are {missing_X.sum()} missing values in X, filling them with the mean")
    X.fillna(X.mean())
else:
    print("There are no missing values in X")

There are no missing values in X


## **Step 3:** Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [8]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

# Import `LinearRegression` from sklearn
from sklearn.linear_model import LinearRegression

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Instantiate model
model = LinearRegression()

# Implement the machine learning model with `X` and `y`
model.fit(X_train, y_train)


LinearRegression()

## **Step 4:** Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [9]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import mean_squared_error, r2_score

# Predictions
training_predictions = model.predict(X_train)
validation_predictions = model.predict(X_test)

# MSE
train_mse = mean_squared_error(y_train, training_predictions)
validation_mse = mean_squared_error(y_test, validation_predictions)

# R2 score
train_r2 = r2_score(y_train, training_predictions)
validation_r2 = r2_score(y_test, validation_predictions)



## **Step 5:** Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [10]:
# TO DO: ADD YOUR CODE HERE
results = pd.DataFrame(columns=["Training Accuracy", "Validation Accuracy"])
results.loc['MSE'] = [train_mse, validation_mse]
results.loc['R2 Score'] = [train_r2, validation_r2]
print(results)

          Training Accuracy  Validation Accuracy
MSE              110.345501            95.635335
R2 Score           0.609071             0.636898


## **Questions (2 marks)**
1. Did using a linear model produce good results for this dataset? Why or why not?

<font color='Green'><b>YOUR ANSWER HERE</b></font>

The use of a linear model seems to produce both good and bad results. The model produced a very high MSE for both the training and validation so that means that the model's predictions were very far away from the actual results. The model produced an R2 score of about 0.60, so the model is able to comprehend around 60% of the entire data.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>Explain YOUR PROCESS here:</b></font>

1. I used the lab assignments as a guideline for my validation metrics code, I also refered to the lecture slides on validation metrics for further insight. The slides also refered me to the online textbook which gave me more examples on validation metric techniques on python.

2. I followed the steps in the order the notebook has specified.

3. I used generative AI to explain how some of the validation metric functions worked, specifically r2_score() and mean_squared_error(). I asked how it worked, what it takes in and what outputs so I could learn how to apply them in the context of the assignment.

4. I mainly had issues learning to use the validation metrics functions as there are not that many examples in our lab assignments and the lectures. I used chatgpt to give me an quick and basic rundown of how they worked which made implementing them much easier.

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


<font color='Green'><b>ADD YOUR FINDINGS HERE</b></font>

A realization I found after finishing the assignment is that the size of the dataset, both feature wise and datapoints wise, greatly impacts the ability for a model to learn from its dataset. More data in general is mostly good, it gives the model more to learn from, as explained in the lectures, I see now why linear models are better on larger datasets. Like in this assignment, part 1 had a much larger dataset with ~4600 than part 2's ~1000, part 2 was much less accurate with abnormaly high MSEs greater than 90. Part 1's dataset also had much more features, with it having 57 compared to part 2's 8 features. If part 2's dataset had more features, as long as theyre relevant, the model would likely be able to make better predictions and lower the MSE and get a higher R2 score.

# **Part 4: Reflection (2 marks)**
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


<font color='Green'><b>ADD YOUR THOUGHTS HERE</b></font>

- I enjoy the python notebook format, much better than having a separate python program and assignment document.
- I also liked how easy it is to relate the assignment to previous lectures and lab assignments.
- I also enjoyed the more "hands-on" experience when applying some of the course theory, makes it easier to understand.
- The only disliked some of the wording in some steps, I had to redo some code after figuring out the big picture of the task.

# **Part 5: Bonus Question (4 marks)**

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [11]:
# TO DO: ADD YOUR CODE HERE
# Imports
from sklearn.linear_model import Ridge, Lasso, RidgeCV, LassoCV

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Make a dict of the models for a loop
alpha_values = [0.001, 0.01, 0.1, 1, 10, 100]
regression_models = {
    'Ridge' : RidgeCV(alphas=alpha_values).fit(X_train, y_train),
    "Lasso" : LassoCV(alphas=alpha_values).fit(X_train, y_train)
}

# dataframe for results
results = pd.DataFrame(columns=["Training Accuracy", "Validation Accuracy"])

# HIT THE LOOOOOOOOOOOP
for name, model in regression_models.items():
    # Predictions
    training_predictions = model.predict(X_train)
    validation_predictions = model.predict(X_test)

    # MSE
    train_mse = mean_squared_error(y_train, training_predictions)
    validation_mse = mean_squared_error(y_test, validation_predictions)

    # R2 score
    train_r2 = r2_score(y_train, training_predictions)
    validation_r2 = r2_score(y_test, validation_predictions)    

    print(f"Best alpha for {name} model: {model.alpha_}")
    results.loc['MSE'] = [train_mse, validation_mse]
    results.loc['R2 Score'] = [train_r2, validation_r2]

    print(results)
    print()


Best alpha for Ridge model: 100.0
          Training Accuracy  Validation Accuracy
MSE              110.345597            95.625173
R2 Score           0.609071             0.636937

Best alpha for Lasso model: 0.001
          Training Accuracy  Validation Accuracy
MSE              110.345501            95.634971
R2 Score           0.609071             0.636899



<font color='Green'><b>ADD YOUR ANSWER HERE</b></font>

Best alpha for Ridge model: 100.0
Best alpha for Lasso model: 0.001

The MSE and R2 scores are not that different from the results obtained from part 2. I explained before how the problem with the linear model's dataset is that there is not enough to learn from, ridge and lasso specialize on regularization, it means that they are better suited to prevent overfitting. The problem with our linear model is that it is suffering from underfitting, so ridge and lasso wont be effective in improving the scores.